### PR TITLE
[solution]:zh/medium-drop-char

### PR DESCRIPTION
**Pull request by martisss**
# Challenge [zh/medium-drop-char]

## Checklist

- [x] Make sure that you are opening the PR with a single solution (one solution per PR)
- [x] The solution must have at least several references for the readers to dive in depth

### CODE REVIEW

**Review by ghaiklor — COMMENTED** [review]
LGTM, waiting for others to approve

File: zh/medium-drop-char.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+```typescript
+type Butterfly = DropChar<" b u t t e r f l y ! ", " ">; // 'butterfly!'
+```
+
+## 解答
+
+为了解决这个问题，我们需要了解模板字面量类型。你可以在`TypeScript Handbook`中[了解更多](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html)。
+
+使用模板字面量类型时，我们可以从字符串中推断所需的部分，并检查其是否是我们期望的部分。让我们从最简单的情况开始 - 推断字符串的左侧部分和右侧部分。它们之间的定界符是所需的字符本身。
```
Comment (likui628) [inline zh/medium-drop-char.md:23]: ‘分隔符’ 比 ‘定界符’ 更常用

File: zh/medium-drop-char.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+使用这样的表示方法，我们会得到一个编译错误`Type ‘C’ is not assignable to type ‘string | number | bigint | boolean | null | undefined’.`添加一个泛型约束去修复它。
+
+```typescript
+type DropChar<S, C extends string> = S extends `${infer L}${C}${infer R}`
+  ? never
+  : never;
+```
+
+现在我们有了左右两部分以及字符`C`。由于我们需要删除`C`，因此我们在没有它的情况下将左右部分返回。
```
Comment (likui628) [inline zh/medium-drop-char.md:37]: 因此我们返回没有它的左右部分。

**Review by likui628 — COMMENTED** [review]
PTAL

**Review by likui628 — APPROVED** [review]

**Pull request merged**